core/src/state/world_state.rs

```diff
@@ -100,8 +142,10 @@ impl WorldState {
         cluster.apply(message.message);
         self.logical_time = sequence;
 
-        if let Some(sender) = self.listeners.remove(&sequence) {
-            sender.notify_waiters();
+        if self.listeners.read().unwrap().contains_key(&sequence) {
```

Comment:
Technically there is a race condition here (albeit an improbable one): if another thread adds the value between the check and the next line, we miss it. Using a `DashMap` will fix this anyway, but in general my inclination is to write-lock for a check-and-set operation (like `remove`) even if the write might not happen.

Reply:
I actually did think of that, but since this happens *after* self.logical_time=sequence, the other thread should return with an activated notify.

I guess it's possible for the other thread to clear the check in get_listener, then get preempted, have this one cross the read lock and then have that thread get the write lock. ugh. I really wanted to avoid having to get a write lock here most of the time since it'll contend hard :(

I'd *really* rather not use DashMap, I wasted like 3 hours debugging weird deadlocks in that library. 

I'll see if there's a better solution, but in the meantime I'll just take a write lock and use the entry API. If/when we need more perf I can switch to doing atomic shenanigans on self.logical_time